torch/_dynamo/output_graph.py

```diff
@@ -219,6 +223,11 @@ def __init__(
         # should use original graphargs.
         self.orig_graphargs: List[GraphArg] = self.graphargs
         self.nn_modules: Optional[Dict[str, torch.nn.Module]] = dict()
+
+        # A map of normalized names (see normalize_attr_name to their sources)
+        # There is a 1:1 relationship between the module attributes written to the root module
+        # and the names in this map.
+        self.nn_modules_sources: Optional[Dict[str, Source]] = dict()
```

Comment:
You can't do this. You are requiring a 1:1 relationship, but there is nothing in `normalize_attr_name` that enforces a 1:1 relationship. For example, let's suppose you have an NN module with parameters `foo_bar` and a submodule `foo` which has a parameter `bar`. `foo_bar` and `foo.bar` normalize to the same string, and will conflict in the map here. That's bad!

Reply:
This does use source name for the nested properties, right?

1) Top level register_attr_or_module, uses top level name
2) Iterate over params, make a source `new_source = ParamBufferSource(source, n)`
3) Invoke register_attr_or_module with the source name from 2

```
new_source = ParamBufferSource(source, n)
new_name = new_source.name()
self.register_attr_or_module(p, new_name, source=new_source)
```

I think the problem, as you pointed out, is that normalize_attr_name breaks this uniqueness that sources give us. Let me see if we can maybe find a better way...

We could also probe self.nn_module_sources for if the key exists, and in that case, increment the name to have _0 on it...